### ISSUE TITLE
Help: Openvpn custom configuration file

### ISSUE THREAD

**Issue by GrumpyOldGeordie** [qdm12/gluetun#223]
**TLDR**: use OpenVPN custom configuration file?

1. Is this urgent?

    - [ ] Yes
    - [x] No

2. What VPN service provider are you using?

    - [ ] PIA
    - [ ] Mullvad
    - [x] Windscribe
    - [ ] Surfshark
    - [ ] Cyberghost

3. What's the version of the program?

Running version latest built on 2020-08-18T01:08:54Z (commit b10a476)

4. What are you using to run the container?

    - [ ] Docker run
    - [x] Docker Compose
    - [ ] Kubernetes
    - [ ] Docker stack
    - [ ] Docker swarm
    - [ ] Podman
    - [x] Other: Portainer 1.23.2 (upgrading to 1.24.1 soon)

5. Extra information
Host OS: OpenMediaVault 5.4.6-1 on Raspian Buster; RPi-4B

Hey there!

Having just come across Gluetun this morning it looks like it could be the answer to a lot of my prayers.  Thanks to all!

I'm intending to build an OpenVPN server on top of a media server to:
1. connect LAN hosts to Windscribe using an http proxy, and
2. provide casual inbound VPN access to the NAS for laptops/phones using a separate OpenVPN instance.

Gluetun gives me both parts of a), the built-in config for Windscribe and Tinyproxy appear to be working fine.  I already have a separate container with OpenVPN configured as an inbound routed VPN server.  As I'm on RPi's, OpenVPN arm images aren't thick on the ground so the inbound VPN is currently on a little-maintained image from "darathor" - ideally I'd like to run both out of separate Gluetun containers.

**Question**: is it possible to get the Gluetun startup to consume a persistent custom openvpn.conf which will reference its own PKI, etc?  I _think_ that's all I'd need to do to run the inbound VPN configured as I need it.

I only got Gluetun running this morning so I'm a complete noob here, but I can't see how to specify a "-config" option to point OpenVPN to use my roll-your-own config.

Cheers, and thanks again,

Keith

**Comment by qdm12** [comment]
Hello! Thanks for the suggestion. It is definitely possible although quite some tweaking would be needed. I'll work on it, but can't tell how soon as I have things to do quite piled up these days 😄

**Comment by GrumpyOldGeordie** [comment]
Great news, thanks.

I'd offer to lend a hand but I haven't written code for over a decade so I'd probably get in the way.  Let me know if you need help with testing though.

Many thanks.

**Comment by qdm12** [comment]
Can you try the image `qmcgaw/gluetun:ovpn-file` with the steps described [in the Wiki here](https://github.com/qdm12/gluetun/wiki/Openvpn-file)?

It works for me, but I was curious about your experience. Let me know if you feel anything is sketchy or should be improved before I merge this in the main codebase. Thanks!

**Comment by GrumpyOldGeordie** [comment]
Hey, great news again!  Can't do that right now but I'll try to make time tomorrow: fingers crossed.

Thanks in anticipation.

**Comment by GrumpyOldGeordie** [comment]
Hi, Quentin,

Didn't get anywhere as a server but the client side is a bit better.

The build for ovpn-file is AMD64 ... my target server is ARM (Raspberry Pi 4B under Debian Buster/OMV5) so I didn't get anywhere with a server config.  When I get more time I'll attempt a server config on a laptop (I've just installed Docker/Portainer to check the client side so I'll build that out a little).

1. As you mentioned in assumed behaviours, you expected a problem with the remote option: static vs. dynamic IP's.  I use DuckDNS to make the server IP available to connecting clients; when the server-side ISP changes the IP, DuckDNS republishes it for the domain so my clients can't use a fixed IP.
2. This Gluetun insists on OPENVPN_USER and OPENVPN_PASSWORD being set.  Rightly or wrongly I'm only using PKI certificates without username and password.  That said duff values appear to safely bypass the errors.
3. Finding the config file appears to require lower-case filenames only (I use mixed-case).  Easily fixed but mixed-case would be nice.
4. I've always used specified remote as "remote XXX.duckdns.org 1194 udp"; Gluetun won't parse three parameters to "remote"; had to explicitly configure "proto udp".
5. Not sure whether this is Gluetun or my local config, but it's not seeing the fact that /dev/net/tun exists on startup and is complaining then creating its own in the container despite it being mapped by the Portainer stack file.

That said, it connects to my routed server and I can ping all hosts on the server LAN from the OpenVPN console shell on the client Portainer.  There are routing or firewall issues preventing applications being accessed ... I'll sort them out when I can; just pinging is a massive step though.

Finally, I also attempted to run a bridged client as I use bridged OpenVPN as a semi-permanent connection between two locations.  Having doctored the bridged client config in line with the above I could see the connection attempt on the server but the client then failed.  With verbosity at the default (the target.conf says it's 1):

2021/03/07 15:13:21 INFO openvpn: /sbin/ip addr add dev tap0 192.168.215.121/24 broadcast 192.168.215.255
2021/03/07 15:13:21 INFO openvpn: Initialization Sequence Completed
2021/03/07 15:13:21 WARN cannot find VPN destination IP address from ip routes
2021/03/07 15:13:21 INFO dns over tls: downloading DNS over TLS cryptographic files
2021/03/07 15:13:21 WARN dns over tls: cannot update files
2021/03/07 15:13:21 INFO dns over tls: attempting restart in 10s

On the warning about VPN destination addresses; there's only one route in the server configuration: to the tunnel network itself.

Too much for a "comment" there: how would you like me to proceed with these?

Thanks,

Keith

**Comment by qdm12** [comment]
> Didn't get anywhere as a server but the client side is a bit better.

Oh yes, it's definitely not designed to be a VPN server, and will likely never be.

> The build for ovpn-file is AMD64

Just changed that in aca112fa42b9f736f6e23cc56e6413892bc5463b so you should have ARM builds now.

For your numbered questions:

1. This may get addressed with an option to allow the DNS resolution of **only** the remote hostname. This will likely come after the #137 which is not there yet unfortunately. There are a few DNS related issues that require me to develop a custom DNS implementation, so that will be batched into it I guess.
2. I disabled the requirement for username and password when using a custom config in 4b6f49b7515270781c20227a56c2657096a6a930
3. Fixed in 6b1891908778eb4a5cbc3a0ae04baa2166a0ec80
4. It now supports both (5ed89c3bd171ed9879fb8f332ce9a7d5d1818b11)
5. Are you bind mounting it? You should use the special `--device` or `devices:` to have it mapped in the container. Although you don't really need it, creating it should not interfere with anything (except the complain message haha).

The firewall is indeed restrictive here and there, although it should allow everything through the tunnel. You can dig into it by exec'ing in the container and see the rules with `iptables -nvL`. Let me know if you find anything specific that should be allowed.

For the final error `WARN cannot find VPN destination IP address from ip routes` I'll investigate and get back to you. Maybe you can give the output of `ip route` from within the container?

I'm a bit busy with various things (wireguard these days), but let's keep this thing moving slowly 😉

**Comment by rucknapucknavitz** [comment]
> @rucknapucknavitz Where do you specify the vpn server port? As far as I know, it's usually `remote <ip> <port>`

I guess that's my problem then! I don't see a port option in my config:

```
remote lu.vpnunlimitedapp.com
proto udp
```

I don't see that they publish which ports are in use, so I'll have to find another way to track that down (unless OpenVPN always uses the same port...).

**Comment by rucknapucknavitz** [comment]
> Can you please check you can also use `remote <ip><port>` with the `proto udp` lines together? It should work in both cases.

Yep, works great.

> Now it should even work if you don't specify the port.

Also works just listing `remote <ip>` without the port and with `proto udp`. Defaults to 1194.


Thanks for both updates!

**Comment by qdm12** [comment]
I'll merge all this in the main gluetun image. There is still #404 left to do but that's blocked for now, feel free to subscribe to it. Cheers!
